### PR TITLE
#17858: Binary servlet changed to return 401 for non logged users

### PR DESCRIPTION
**Pull request by dsolistorres**
Before this change, binary servlet returns 404 (Not Found) for a front end user if the requested file asset is not accessible for the user doing the request because permissions are required. Now, the servlet returns 401 (Authorization required), so the user can be redirected to the login page.

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    private static final String BY_ID = "by-identifier";
+    private static final String BY_INODE = "by-inode";
+
+    private static final String READ_PERMISSIONS = "has-read-permissions";
+    private static final String NO_PERMISSIONS = "no-permissions";
+
+    private static Host host = null;
+    private static Role role = null;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [Avoid using redundant field initializer for 'role'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        return this;
+    }
+
+    @Override
+    public void sendError(int status) throws IOException {
+        this.status = status;
+    }
+
+    @Override
+    public void setStatus(int status) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java]: Codacy Issue found: [Parameter 'status' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        }
+    }
+
+    private static final String BY_ID = "by-identifier";
+    private static final String BY_INODE = "by-inode";
+
+    private static final String READ_PERMISSIONS = "has-read-permissions";
+    private static final String NO_PERMISSIONS = "no-permissions";
+
+    private static Host host = null;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [Avoid using redundant field initializer for 'host'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    private static Role role = null;
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        // Set testing environment
+        IntegrationTestInitService.getInstance().init();
+
+        host = new SiteDataGen().nextPersisted();
+        role = new RoleDataGen().nextPersisted();
+    }
+
+    @DataProvider
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [JUnit 4 tests that execute tests should use the @Test annotation, JUnit 5 tests should use @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+@RunWith(DataProviderRunner.class)
+public class BinaryExporterServletTest {
+
+    // Temporary binary png file
+    private static class TmpBinaryFile implements Closeable {
+
+        private final Path pngFilePath;
+
+        public TmpBinaryFile(boolean setContent) throws IOException {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [Parameter 'setContent' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.File;
+import java.io.IOException;
+
+public class MockHttpStatusResponse extends HttpServletResponseWrapper implements MockResponse {
+
+    private int status = HttpServletResponse.SC_OK;
+
+    public MockHttpStatusResponse(HttpServletResponse response) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java]: Codacy Issue found: [Parameter 'response' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+    }
+
+    private HttpServletRequest mockServletRequest(final String fileURI) {
+        return new MockSessionRequest(new MockServletPathRequest(
+                new MockHttpRequest("localhost", fileURI).request(),
+                "/contentAsset"));
+    }
+
+    private HttpServletResponse mockServletResponse(TmpBinaryFile tmpTargetFile) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [Parameter 'tmpTargetFile' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                { BY_INODE, READ_PERMISSIONS },
+                { BY_INODE, NO_PERMISSIONS },
+        };
+    }
+
+
+    @Test
+    @UseDataProvider("testCases")
+    public void requestBinaryFile(
+            final String byIdType, final String permissionType) throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        super(response);
+    }
+
+    @Override
+    public HttpServletResponse response() {
+        return this;
+    }
+
+    @Override
+    public void sendError(int status) throws IOException {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java]: Codacy Issue found: [Parameter 'status' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    private void sendRequest(final HttpServletRequest request, final HttpServletResponse response)
+            throws ServletException, IOException {
+
+        final BinaryExporterServlet binaryExporterServlet = new BinaryExporterServlet();
+        binaryExporterServlet.init();
+        binaryExporterServlet.doGet(request, response);
+
+    }
+
+    private void addPermissions(Contentlet fileAsset)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [Parameter 'fileAsset' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package com.dotcms.mock.response;
+
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.File;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java]: Codacy Issue found: [Avoid unused imports such as 'java.io.File'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+    final private boolean byIdentifier;
+    final private boolean permissionsRequired;
+
+    public BinaryExporterServletTest(final String byIdType, final String permissionType) {
+        this.byIdentifier = byIdType.equals(BY_ID);
+        this.permissionsRequired = permissionType.equals(NO_PERMISSIONS);
+    }
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+        // Set testing environment
+        IntegrationTestInitService.getInstance().init();
+
+        host = new SiteDataGen().nextPersisted();
+        role = new RoleDataGen().nextPersisted();
+    }
+
+    @Parameters(name = "{0},{1}")
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [JUnit 4 tests that execute tests should use the @Test annotation, JUnit 5 tests should use @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.io.IOException;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+public class MockHttpStatusResponse extends HttpServletResponseWrapper implements MockResponse {
+
+    private int status = HttpServletResponse.SC_OK;
+
+    public MockHttpStatusResponse(HttpServletResponse response) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/mock/response/MockHttpStatusResponse.java]: Codacy Issue found: [Parameter 'response' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+
+        // Set testing environment
+        IntegrationTestInitService.getInstance().init();
+
+        host = new SiteDataGen().nextPersisted();
+        role = new RoleDataGen().nextPersisted();
+
+    }
+
+    @DataProvider
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [JUnit 4 tests that execute tests should use the @Test annotation, JUnit 5 tests should use @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

File: dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    private static final String BY_INODE = "by-inode";
+
+    private static final String READ_PERMISSIONS = "has-read-permissions";
+    private static final String NO_PERMISSIONS = "no-permissions";
+
+    private static Host host;
+    private static Role role;
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/servlets/BinaryExporterServletTest.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4885343)

**Review by dsilvam — APPROVED** [review]

**Review by jgambarios — APPROVED** [review]

**Pull request merged**